### PR TITLE
630: Only nack messages upon cancellation of a consumer subscription …

### PR DESCRIPTION
**Pull request by lfse-slafleur**
…that belong to that consumer subscription.

### CODE REVIEW

**Comment by lfse-slafleur** [comment]
Closes #630

File: tests/test_amqp.py [hunk 1]
```diff
@@ -1587,6 +1587,141 @@ async def test_heartbeat_disabling(
         async with connection:
             assert heartbeat == 0
 
+    async def test_non_acked_messages_are_redelivered_to_queue(
```
Comment (lfse-slafleur) [inline tests/test_amqp.py:1590]: Test to see if messages are properly nack'ed and I didn't break anything.

File: tests/test_amqp.py [hunk 2]
```diff
@@ -1587,6 +1587,141 @@ async def test_heartbeat_disabling(
...
+
+            assert incoming_message.body == all_bodies[i]
+
+        # Check if the queue is now empty
+        assert await queue.get(fail=False, timeout=.5) is None
+
+        # Cleanup, delete the queue
+        await queue.delete()
+
+    async def test_regression_only_messages_cancelled_subscription_are_nacked(
```
Comment (lfse-slafleur) [inline tests/test_amqp.py:1636]: Regression test which failed before fixing the code. Documentation is inline to show why the test is constructed the way it is.

File: aio_pika/queue.py [hunk 3]
```diff
@@ -441,32 +441,24 @@ async def close(self, *_: Any) -> Any:
...
 
-        # Reject all messages
-        msg: Optional[IncomingMessage] = None
-        try:
-            while True:
-                msg = self._queue.get_nowait()
-        except asyncio.QueueEmpty:
-            if msg is None:
-                return
+        # Reject all messages that have been received and in the buffer/cache.
```
Comment (lfse-slafleur) [inline aio_pika/queue.py:444]: While I fixed the code here locally as I thought it makes sense, something does still is uneasy with me. I would expect that the code here is not necessary and may even contain a liveliness issue.

Regarding the liveliness issue: Assume a queue that is receiving messages rapidly. The 'while' loop would be stuck manually nack'ing each and every single message until it is fast enough to escape the 'while' loop. It is possible this never happens. Also, it is possible that messages are still received after escaping the 'while' loop but just before the subscription termination message is send to RabbitMQ. These messages are not nack'ed.

Regarding the necessity: If an application wants each message to arrive, the message would not be automatically ack'ed but rather manually ack'ed after the message is persisted by whatever application integrates aio-pika. If a different delivery guarantee is expected, I would expect the application to allow for messages being dropped or duplicated. In other words, if the local storage is not explicitely ack'ed, the whole system should still work 'as expected'. ('As expected' here means as configured. If is a message is auto-acked (at-most-once guarantee), the application should be able to deal with missing messages. If at-least-once or exactly-once is expected, RabbitMQ should just requeue the moment the connection is dropped even if we do not requeue the messages).

So I am a bit confused as to why this code is required. @mosquito  Perhaps you could elaborate / provide guidance? Thanks in advance!

**Review by mosquito — APPROVED** [review]

**Comment by mosquito** [comment]
@lfse-slafleur released as 9.4.2

**Pull request merged**